git-branchless-lib/src/core/rewrite/plan.rs

```diff
@@ -1056,4 +1156,594 @@ mod tests {
+            builder.move_commit(test3_oid, test1_oid)?;
+            builder.move_commit(test5_oid, test3_oid)?;
+            Ok(())
+        })?;
+
+        let (stdout, _stderr) = git.run(&["smartlog"])?;
+        insta::assert_snapshot!(stdout, @r###"
+        O f777ecc (master) create initial.txt
+        |
+        o 62fc20d create test1.txt
+        |\
+        | o ad2c2fc create test3.txt
+        | |
+        | @ ee4aebf create test5.txt
+        |
+        o 96d1c37 create test2.txt
+        |
+        o 8556cef create test4.txt
+        "###);
+
+        Ok(())
+    }
+
+    #[test]
+    fn test_plan_moving_range_again_overrides_previous_move() -> eyre::Result<()> {
+        let git = make_git()?;
+        git.init_repo()?;
+        git.detach_head()?;
+        let test1_oid = git.commit_file("test1", 1)?;
+        let test2_oid = git.commit_file("test2", 2)?;
+        git.commit_file("test3", 3)?;
+        let test4_oid = git.commit_file("test4", 4)?;
+        let test5_oid = git.commit_file("test5", 5)?;
+        git.commit_file("test6", 6)?;
+
+        create_and_execute_plan(&git, move |builder: &mut RebasePlanBuilder| {
+            builder.move_range(test4_oid, test5_oid, test1_oid)?;
+            builder.move_range(test4_oid, test5_oid, test2_oid)?;
+            Ok(())
+        })?;
+
+        let (stdout, _stderr) = git.run(&["smartlog"])?;
+        insta::assert_snapshot!(stdout, @r###"
+        O f777ecc (master) create initial.txt
+        |
+        o 62fc20d create test1.txt
+        |
+        o 96d1c37 create test2.txt
+        |\
+        | o 8556cef create test4.txt
+        | |
+        | o 566236a create test5.txt
+        |
+        o 70deb1e create test3.txt
+        |
+        @ 35928ae create test6.txt
+        "###);
+
+        Ok(())
+    }
+
+    #[test]
+    fn test_plan_moving_range_and_then_parital_beginning_range_again() -> eyre::Result<()> {
+        let git = make_git()?;
+        git.init_repo()?;
+        git.detach_head()?;
+        let test1_oid = git.commit_file("test1", 1)?;
+        git.commit_file("test2", 2)?;
+        let test3_oid = git.commit_file("test3", 3)?;
+        let test4_oid = git.commit_file("test4", 4)?;
+        let test5_oid = git.commit_file("test5", 5)?;
+        git.commit_file("test6", 6)?;
+
+        create_and_execute_plan(&git, move |builder: &mut RebasePlanBuilder| {
+            builder.move_range(test3_oid, test5_oid, test1_oid)?;
+            builder.move_range(test3_oid, test4_oid, test1_oid)?;
+            Ok(())
+        })?;
+
+        let (stdout, _stderr) = git.run(&["smartlog"])?;
+
+        // FIXME This output is correct mathematically, but feels like it should
+        // be incorrect. What *should* we be doing if the user moves 2 ranges w/
+        // the same source/root oid but different end oids??
+        //
+        // NOTE See also the next test for the other case: where the end of the
+        // range is moved again. That *does* feel correct.
```

Comment:
I want to bring you attention to this case/test. I understand why it's correct, but it feels a little surprising. ATM there's no way (that I know) to handle it any differently.

Reply:
This looks fine to me as-is. There are a few options:

- Produce an error in this case.
- Convert these commits to be merge commits (multi-parent commits). It would probably be better to to make `move_range` accept a set of end OIDs, rather than have the caller call the same function multiple times.